src/EditorFeatures/CSharpTest/GenerateVariable/GenerateVariableTests.cs

```diff
@@ -7607,6 +7607,368 @@ void Method()
+        public async Task TestGenerateFieldInExpressionBodiedLocalFunctionInferredFromType()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Program
+{
+    public int Y()
+    {
+        return Foo();
+
+        int Foo() => [|y|] = 12;
+    }
+}",
+@"class Program
+{
+    private int y;
+
+    public int Y()
+    {
+        return Foo();
+
+        int Foo() => y = 12;
+    }
+}");
+        }
+
+        [WorkItem(26993, "https://github.com/dotnet/roslyn/issues/26993")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateVariable)]
+        public async Task TestGenerateFieldInBlockBodiedLocalFunction()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Program
+{
+    public int Y()
+    {
+        return Foo();
+
+        int Foo()
+        {
+            return [|y|];
+        }
+    }
+}",
+@"class Program
+{
+    private int y;
+
+    public int Y()
+    {
+        return Foo();
+
+        int Foo()
+        {
+            return y;
+        }
+    }
+}");
+        }
+
+        [WorkItem(26993, "https://github.com/dotnet/roslyn/issues/26993")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateVariable)]
+        public async Task TestGenerateReadonlyFieldInBlockBodiedLocalFunction()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Program
+{
+    public int Y()
+    {
+        return Foo();
+
+        int Foo()
+        {
+            return [|y|];
+        }
+    }
+}",
+@"class Program
+{
+    private readonly int y;
+
+    public int Y()
+    {
+        return Foo();
+
+        int Foo()
+        {
+            return y;
+        }
+    }
+}",
+index: 1);
+        }
+
+        [WorkItem(26993, "https://github.com/dotnet/roslyn/issues/26993")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateVariable)]
+        public async Task TestGeneratePropertyInBlockBodiedLocalFunction()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Program
+{
+    public int Y()
+    {
+        return Foo();
+
+        int Foo()
+        {
+            return [|y|];
+        }
+    }
+}",
+@"class Program
+{
+    public int y { get; private set; }
+
+    public int Y()
+    {
+        return Foo();
+
+        int Foo()
+        {
+            return y;
+        }
+    }
+}",
+index: 2);
+        }
+
+        [WorkItem(26993, "https://github.com/dotnet/roslyn/issues/26993")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateVariable)]
+        public async Task TestGenerateFieldInBlockBodiedLocalFunctionInferredFromType()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Program
+{
+    public int Y()
+    {
+        return Foo();
+
+        int Foo() 
+        {
+            return [|y|] = 12;
+        }
+    }
+}",
+@"class Program
+{
+    private int y;
+
+    public int Y()
+    {
+        return Foo();
+
+        int Foo() 
+        {
+            return y = 12;
```

Comment:
I think this isn't fixed. If you change the return type of `Y` to void, this won't work. It's still working incorrectly by taking the return type of the containing method, not this local function.

Reply:
Just try this simple case:
```c#
class C
{
    void M()
    {
        int F()
        {
            return y;
        }
    }
}
```
the generated field will be of type object.

Reply:
Also, I think we shouldn't use `Foo` as a name although I can't tell you why.

Reply:
You're right regarding the block bodied types. I think I'll include that too in this pull request, might take me up to a week though because I'm busy for a while. I'll tag you once it's done.